clang/runtime/dpct-rt/include/dpct/blas_utils.hpp

```diff
@@ -2467,8 +2467,618 @@ trmm(sycl::queue &q, oneapi::mkl::side left_right,
+      CASE(sm_count_target)
+      CASE(a_scale_pointer)
+      CASE(b_scale_pointer)
+      CASE(c_scale_pointer)
+      CASE(d_scale_pointer)
+      CASE(amax_d_pointer)
+      CASE(epilogue_aux_data_type)
+      CASE(epilogue_aux_scale_pointer)
+      CASE(epilogue_aux_amax_pointer)
+      CASE(fast_accum)
+      CASE(bias_data_type)
+      CASE(atomic_sync_in_counters_pointer)
+      CASE(atomic_sync_out_counters_pointer)
+      CASE(atomic_sync_num_chunks_d_rows)
+      CASE(atomic_sync_num_chunks_d_cols)
+    }
+#undef CASE
+  }
+
+  compute_type _compute_type;
+  library_data_t _scale_type;
+  pointer_mode_t _pointer_mode = pointer_mode_t::host;
+  oneapi::mkl::transpose _trans_a = oneapi::mkl::transpose::nontrans;
+  oneapi::mkl::transpose _trans_b = oneapi::mkl::transpose::nontrans;
+  oneapi::mkl::transpose _trans_c = oneapi::mkl::transpose::nontrans;
+  oneapi::mkl::uplo _fill_mode;
+  int _epilogue = 1;
+  void *_bias_pointer;
+  std::int64_t _bias_batch_stride;
+  void *_epilogue_aux_pointer;
+  std::int64_t _epilogue_aux_ld;
+  std::int64_t _epilogue_aux_batch_stride;
+  std::int64_t _alpha_vector_batch_stride;
+  std::int32_t _sm_count_target;
+  const void *_a_scale_pointer;
+  const void *_b_scale_pointer;
+  const void *_c_scale_pointer;
+  const void *_d_scale_pointer;
+  void *_amax_d_pointer;
+  library_data_t _epilogue_aux_data_type;
+  void *_epilogue_aux_scale_pointer;
+  void *_epilogue_aux_amax_pointer;
+  std::int8_t _fast_accum;
+  library_data_t _bias_data_type;
+  std::int32_t *_atomic_sync_in_counters_pointer;
+  std::int32_t *_atomic_sync_out_counters_pointer;
+  std::int32_t _atomic_sync_num_chunks_d_rows;
+  std::int32_t _atomic_sync_num_chunks_d_cols;
+
+  friend void matmul(matmul_desc_ptr computeDesc, const void *alpha,
+                     const void *a, matrix_layout_ptr a_desc, const void *b,
+                     matrix_layout_ptr b_desc, const void *beta, const void *c,
+                     matrix_layout_ptr c_desc, void *d,
+                     matrix_layout_ptr d_desc, dpct::queue_ptr q_ptr);
+};
+
+namespace detail {
+dnnl::memory::data_type dpct_type_to_dnnl_type(library_data_t type) {
+  switch (type) {
+  case library_data_t::real_half:
+    return dnnl::memory::data_type::f16;
+  case library_data_t::real_bfloat16:
+    return dnnl::memory::data_type::bf16;
+  case library_data_t::real_float:
+    return dnnl::memory::data_type::f32;
+  case library_data_t::real_double:
+    return dnnl::memory::data_type::f64;
+  case library_data_t::real_int32:
+    return dnnl::memory::data_type::s32;
+  case library_data_t::real_int8:
+    return dnnl::memory::data_type::s8;
+  case library_data_t::real_uint8:
+    return dnnl::memory::data_type::u8;
+  default:
+    throw std::runtime_error(
+        "the input library_data_t type has no corresponding type in dnnl");
+  }
+}
+} // namespace detail
+
+void matmul(matmul_desc_ptr compute_desc, const void *alpha, const void *a,
```

Comment:
a/b: convert to col major layout, allocate new memory for col major layout d to save result, then do the calculation.
After calculation, convert col major layout d to col32 layout.

Reply:
done